tangelo/toolboxes/operators/operators.py

```diff
@@ -21,16 +21,89 @@
 
 import numpy as np
 from scipy.special import comb
+import openfermion as of
 
-# Later on, if needed, we can extract the code for the operators themselves to remove the dependencies and customize
-import openfermion
 
-
-class FermionOperator(openfermion.FermionOperator):
-    """Currently, this class is coming from openfermion. Can be later on be
-    replaced by our own implementation.
+class FermionOperator(of.FermionOperator):
+    """Custom FermionOperator class. Based on openfermion's, with additional functionalities.
     """
 
+    def __init__(self, term=None, coefficient=1., n_spinorbitals=None, n_electrons=None, spin=None):
+        super(FermionOperator, self).__init__(term, coefficient)
+
+        self.n_spinorbitals = n_spinorbitals
+        self.n_electrons = n_electrons
+        self.spin = spin
+
+    def __imul__(self, other):
+        if isinstance(other, FermionOperator):
+            # Raise error if attributes are not the same across Operators.
+            if (self.n_spinorbitals, self.n_electrons, self.spin) != (other.n_spinorbitals, other.n_electrons, other.spin):
+                raise RuntimeError("n_spinorbitals, n_electrons and spin must be the same for all FermionOperators.")
+            else:
+                return super(FermionOperator, self).__imul__(other)
+
+        elif isinstance(other, of.FermionOperator):
+            if (self.n_spinorbitals, self.n_electrons, self.spin) != (None, None, None):
+                raise RuntimeError("openfermion FermionOperator did not define a necessary attribute")
+            else:
+                f_op = FermionOperator()
+                f_op.terms = other.terms.copy()
+                return super(FermionOperator, self).__imul__(f_op)
+
+        else:
+            return super(FermionOperator, self).__imul__(other)
+
+    def __mul__(self, other):
+        return self.__imul__(other)
+
+    def __rmul__(self, other):
+        return self.__imul__(other)
```

Comment:
I don't think you need to redefine this. From openfermion `SymbolicOperator`

```python3
    def __rmul__(self, multiplier):
        """
        Return multiplier * self for a scalar.

        We only define __rmul__ for scalars because the left multiply
        exist for  SymbolicOperator and left multiply
        is also queried as the default behavior.

        Args:
          multiplier: A scalar to multiply by.

        Returns:
          product: A new instance of SymbolicOperator.

        Raises:
          TypeError: Object of invalid type cannot multiply SymbolicOperator.
        """
        if not isinstance(multiplier, COEFFICIENT_TYPES):
            raise TypeError('Object of invalid type cannot multiply with ' +
                            type(self) + '.')
        return self * multiplier
```

Reply:
good point